File: 3615.longest-palindromic-path-in-graph.go
#
# @lc app=leetcode id=3615 lang=golang
#
# [3615] Longest Palindromic Path in Graph
#

# @lc code=start
func maxLen(n int, edges [][]int, label string) int {
	adj := make([][]int, n)
	for _, e := range edges {
		u, v := e[0], e[1]
		adj[u] = append(adj[u], v)
		adj[v] = append(adj[v], u)
	}

	N := 1 << n
	memo := make([]int, n*n*N)

	maxx := func(a, b int) int {
		if a > b {
			return a
		}
		return b
	}

	var dfs func(int, int, int) int
	dfs = func(u, v, mask int) int {
		idx := (u*n + v)*N + mask
		if memo[idx] > 0 {
			return memo[idx]
		}
		res := 1
		for _, nu := range adj[u] {
			if (mask & (1 << nu)) != 0 {
				continue
			}
			lch := label[nu]
			for _, nv := range adj[v] {
				if (mask & (1 << nv)) != 0 {
					continue
				}
				if label[nv] != lch {
					continue
				}
				if nu == nv {
					continue
				}
				newmask := mask | (1 << nu) | (1 << nv)
				cand := 1 + dfs(nu, nv, newmask)
				if cand > res {
					res = cand
				}
			}
		}
		memo[idx] = res
		return res
	}

	ans := 1

	// odd centers
	for c := 0; c < n; c++ {
		cmask := 1 << c
		for _, l := range adj[c] {
			for _, r := range adj[c] {
				if l == r {
					continue
				}
				if label[l] != label[r] {
					continue
				}
				initmask := cmask | (1 << l) | (1 << r)
				arm := dfs(l, r, initmask)
				total := 2*arm + 1
				if total > ans {
					ans = total
				}
			}
		}
	}

	// even centers
	for _, e := range edges {
		for k := 0; k < 2; k++ {
			a := e[k]
			b := e[1 - k]
			if label[a] != label[b] {
				continue
			}
			ans = maxx(ans, 2)
			amask := (1 << a) | (1 << b)
			for _, l := range adj[a] {
				if l == b {
					continue
				}
				for _, r := range adj[b] {
					if r == a {
						continue
					}
					if l == r {
						continue
					}
					if label[l] != label[r] {
						continue
					}
					initmask := amask | (1 << l) | (1 << r)
					arm := dfs(l, r, initmask)
					total := 2 + 2*arm
					if total > ans {
						ans = total
					}
				}
			}
		}
	}

	return ans
}
# @lc code=end

File: 3615.longest-palindromic-path-in-graph.java
#
# @lc app=leetcode id=3615 lang=java
#
# [3615] Longest Palindromic Path in Graph
#

# @lc code=start
class Solution {
    public int maxLen(int n, int[][] edges, String label) {
        boolean[][] graph = new boolean[n][n];
        for (int[] e : edges) {
            graph[e[0]][e[1]] = true;
            graph[e[1]][e[0]] = true;
        }
        final char[] chs = label.toCharArray();
        final int max_mask = 1 << n;
        int[] pop = new int[max_mask];
        for (int i = 0; i < max_mask; i++) {
            pop[i] = Integer.bitCount(i);
        }
        final int[] popc = pop;
        final boolean[][] fgraph = graph;
        final char[] fchs = chs;
        final int NODES = n;
        final int[] mem_temp = new int[16384 * 14 * 14];
        final int[] mem = mem_temp;

        class Helper {
            int dfs(int mask, int l, int r) {
                int idx = (mask * 14 + l) * 14 + r;
                if (mem[idx] > 0) {
                    return mem[idx];
                }
                int res = popc[mask];
                for (int nl = 0; nl < NODES; nl++) {
                    if (!fgraph[l][nl] || (mask & (1 << nl)) != 0) continue;
                    char cl = fchs[nl];
                    for (int nr = 0; nr < NODES; nr++) {
                        if (!fgraph[r][nr] || (mask & (1 << nr)) != 0 || nl == nr || fchs[nr] != cl) continue;
                        int nmask = mask | (1 << nl) | (1 << nr);
                        res = Math.max(res, dfs(nmask, nl, nr));
                    }
                }
                mem[idx] = res;
                return res;
            }
        }
        Helper helper = new Helper();
        int ans = 1;
        // odd centers
        for (int c = 0; c < n; c++) {
            ans = Math.max(ans, helper.dfs(1 << c, c, c));
        }
        // even centers
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (fgraph[i][j] && fchs[i] == fchs[j]) {
                    int msk = (1 << i) | (1 << j);
                    ans = Math.max(ans, helper.dfs(msk, i, j));
                }
            }
        }
        return ans;
    }
}
# @lc code=end